### PR TITLE
Protobuf fixes for Bazel 0.29

### LINKED ISSUE #287: Unable to depend on WKTs with Bazel 0.29
When running Bazel 0.29.0rc6 + `rules_swift` `master`, the following rule fails:
```python
swift_proto_library(
    name = "duration_swift",
    deps = ["@com_google_protobuf//:duration_proto"],
)
```

```
$ bazel build //:duration_swift --verbose_failures
ERROR: /private/var/tmp/_bazel_steeve/435e7d592a4c50f40fe16990c7f36232/external/com_google_protobuf/BUILD:288:2: Generating Swift sources for @com_google_protobuf//:duration_proto failed (Exit 1) mkdir_and_run failed: error executing command
  (cd /private/var/tmp/_bazel_steeve/435e7d592a4c50f40fe16990c7f36232/sandbox/darwin-sandbox/232/execroot/co_znly_protodefs && \
  exec env - \
  bazel-out/host/bin/external/build_bazel_rules_swift/tools/mkdir_and_run/mkdir_and_run bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/duration_proto.protoc_gen_pb_swift bazel-out/host/bin/external/com_google_protobuf/protoc '--plugin=protoc-gen-swift=bazel-out/host/bin/external/com_github_apple_swift_protobuf/ProtoCompilerPlugin' '--swift_out=bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/duration_proto.protoc_gen_pb_swift' '--swift_opt=FileNaming=FullPath' '--swift_opt=Visibility=Public' '--swift_opt=ProtoPathModuleMappings=bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/<target @com_google_protobuf/:duration_proto>.protoc_gen_swift_modules.asciipb' --descriptor_set_in bazel-out/darwin-fastbuild/bin/external/com_google_protobuf/duration_proto-descriptor-set.proto.bin _virtual_imports/duration_proto/google/protobuf/duration.proto)
Execution platform: @bazel_tools//platforms:host_platform

Use --sandbox_debug to see verbose messages from the sandbox
_virtual_imports/duration_proto/google/protobuf/duration.proto: No such file or directory
Target //tileserver:duration_swift failed to build
```

Building with Bazel `0.28.1` works. Leading me to think https://github.com/bazelbuild/bazel/issues/7157 may be the culprit.

### PR DESCRIPTION
**Pull request by steeve**
Fixes #287 

The fix was stolen from various patches found in https://github.com/bazelbuild/bazel/issues/7157

### CODE REVIEW

**Comment by steeve** [comment]
It might be a good idea to backport this on 0.12, since `rules_swift` `master` is not compatible with the latest `swift-protobuf` release because of https://github.com/bazelbuild/rules_swift/commit/8d61a4169953650e351321c610324af41d0b1bf3

File: swift/internal/swift_protoc_gen_aspect.bzl [hunk 1]
```diff
@@ -270,7 +270,7 @@ def _swift_protoc_gen_aspect_impl(target, aspect_ctx):
         minimal_module_mappings.extend(_gather_transitive_module_mappings(proto_deps))
 
     transitive_module_mapping_file = register_module_mapping_write_action(
-        target,
+        target.label.name,
```
Comment (thomasvl) [inline swift/internal/swift_protoc_gen_aspect.bzl]: Why was this needed?

Comment (steeve) [reply swift/internal/swift_protoc_gen_aspect.bzl]: This is not needed for Bazel 0.29, but I figure I would take the change and fix it anyway, that's why it's in its own commit
```
register_module_mapping_write_action expects a string, not a target
```

The bug was that the file was named `path/to/the/file/<generated //xxxxx>.yyy`

**Comment by steeve** [comment]
Also if you want I can change the commit/PR to be:
```
Add support for --incompatible_generated_protos_in_virtual_imports
```

Since things have changed.

File: swift/internal/utils.bzl [hunk 2]
```diff
@@ -232,5 +232,11 @@ def workspace_relative_path(file):
     Returns:
         The path of the file relative to its workspace.
     """
+    # Bazel creates symlinks to the .proto files under a directory called
+    # "_virtual_imports/<rule name>" if we do any sort of munging of import
+    # paths (e.g. using strip_import_prefix / import_prefix attributes)
+    virtual_imports = "/_virtual_imports/"
+    if virtual_imports in file.path:
+        return file.path.split(virtual_imports)[1].split("/", 1)[1]
```
[the remark this answers is not in the document]
Comment (steeve) [reply swift/internal/utils.bzl]: I'm not sure:
> When this flag is flipped, the two .proto files are both symlinked into a directory bazel-bin/x/virtual_imports/p) and that's what proto_source_root will be.

Also, the fix is from:
- https://github.com/grpc/grpc-java/pull/5959/files
- https://github.com/bazelbuild/rules_rust/pull/239/files
So it would mean they are not ok ?

[the remark this answers is not in the document]
Comment (thomasvl) [reply swift/internal/utils.bzl]: From what is being said, yea, they also won't handle a generated proto file being used (unless they are hardcoding other output paths into the proto search path for their protoc invokes)

**Comment by thomasvl** [comment]
I believe #298 covers this using the data from the provider vs. hardcoding things.
